lib/EarlyFunctionAnalysis/UsedRegisters.h

```diff
@@ -0,0 +1,152 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <map>
+#include <set>
+
+#include "revng/EarlyFunctionAnalysis/CallGraph.h"
+#include "revng/EarlyFunctionAnalysis/Common.h"
+#include "revng/MFP/SetLattices.h"
+#include "revng/Model/Register.h"
+#include "revng/Support/MetaAddress.h"
+
+#include "DetectABI.h"
+
+namespace efa {
+
+struct FunctionABI {
+  using SetOfRegisters = std::set<model::Register::Values>;
+  using SUL = SetUnionLattice<SetOfRegisters>;
+  SetOfRegisters ArgumentRegisters;
+  SetOfRegisters ReturnRegisters;
+};
+
+// using LatticeElement = std::map<MetaAddress, FunctionABI>;
+
+struct LatticeElement : std::map<MetaAddress, FunctionABI> {
+  virtual ~LatticeElement() = default;
+
+  [[nodiscard]] virtual bool
+  hasRetReg(MetaAddress Addr, model::Register::Values V) const {
+    if (count(Addr)) {
+      return at(Addr).ReturnRegisters.count(V);
+    }
```

Comment:
Here you're performing a lookup twice.
It's annoying the first time to do the right thing, but once you get used to it it's a no brainer:

```suggestion
    auto It = find(Addr);
    if (It != end()) {
      return It->ReturnRegisters.count(V);
    }
```